marlowe-runtime/docs/filtered-chain-sync-spec-1.0.md

```diff
@@ -0,0 +1,375 @@
+- client: The role played by the participant which sends queries and receives chain data.
+- server: The role played by the participant which responds to queries maintains chain data.
+- state: A member of the set of states specified by the state machine of the protocol.
+- message: A member of the set of state transitions specified by the state maching of the protocol.
+- payload: Data associated and sent with a message.
+- query: Part of the payload of a query next message which the client sends to the server to tell it what to send next.
+- error: A protocol level error which is triggered when one party sends an invalid message. Results in the termination of the connection.
+- rejection: An application level error which is triggered when the server rejects a syntactically valid, semantically invalid query sent by the client. Does not result in the termination of the connection.
+- roll forward: A message sent by the server to the client in response to a query.
+- result: Part of the payload of a roll forward message which contains the result of a query.
+- roll backward: A messagge sent by the server to the client when the client's position has been rolled back.
+- tip: The latest chronological point in the blockchain
+- point: A point in the blockchain up to and including the tip
+- position: The point where a client is currently "focused."
+- schema version: A named and numberedversion of the query schema supported by the parties of the protocol. Used in the handshake to verify that the server supports the query schema supported by the client.
+- handshake: A proceedure used to establish a connection which ensures the client and server understand the same queeries.
+- agency: Used to describe which role is permited to send a message in a given state.
+
+### Operation
+
+The Filtered Chain Sync Protocol is a stateful client-server protocol. When a
+client establishes a connection with a server, a handshake is performed. If the
+handshake succeeds, the idle state is reached. From the idle state, the client
+has the agency to send a query or terminate the connection. If the client
+terminates the connection, by sending a `Done` message, then the server closes
+the connection and the protocol terminates. If the client sends a query, by
+sending a `QueryNext` message, the server can choose to respond immediately to
+the query, reject the query with a rejection reason, or ask the client to wait
+for a response. When the client receives a response or a rejection, it returns
+to the idle state, and the loop repeats.
+
+A response can be a roll forward or roll backward message. If the response is a
+roll forward, the payload will include the new position of the client and the
+result of the query. If the response is a roll backward message, the payload
+will include the new position of the client, and the new tip of the blockchain.
+The position will be the most recent common ancestor point of the client's
+previous position and the new tip.
+
```

Comment:
Also, what happens when the server starts to shut down. Does it notify clients, or just drop the connnections?

Reply:
Hmm, good question. I think it needs to sever the transport-layer connection. The problem is that the server cannot notify clients in the `Idle` state because the client has agency (i.e. it will not be listening for messages on its end of the channel). It could, in theory, send `Closed` messages to all clients waiting for a response, but that doesn't cover all clients. So I think the appropriate action is to terminate the connection at the transport layer.